core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationOutputProcessor.java

```diff
@@ -0,0 +1,149 @@
+    {
+        this.aggregationInputTypes = ImmutableList.copyOf(aggregationInputTypes);
+
+        this.groupedAggregators = requireNonNull(aggregatorFactories, "aggregatorFactories is null")
+                .stream()
+                .map(AggregatorFactory::createGroupedAggregator)
+                .collect(toImmutableList());
+        this.aggregationInputChannels = Ints.toArray(aggregationInputChannels);
+        this.maskBlockChannels = Ints.toArray(maskBlockChannels);
+        this.hashChannels = new int[groupByChannels.size() + (inputHashChannel.isPresent() ? 1 : 0)];
+        for (int i = 0; i < groupByChannels.size(); i++) {
+            hashChannels[i] = groupByChannels.get(i);
+        }
+        inputHashChannel.ifPresent(channelIndex -> hashChannels[groupByChannels.size()] = channelIndex);
+    }
+
+    public Page processAggregatedPage(Page page)
+    {
+        Block[] finalPage = new Block[page.getChannelCount() + maskBlockChannels.length + aggregationInputTypes.size() + 1];
+        for (int i = 0; i < page.getChannelCount(); i++) {
+            finalPage[i] = page.getBlock(i);
+        }
+        int positionCount = page.getPositionCount();
+
+        // mask channels
+        for (int i = 0; i < maskBlockChannels.length; i++) {
+            finalPage[page.getChannelCount() + i] = nullRle(BOOLEAN, positionCount);
+        }
+        // aggregation inputs
+        for (int i = 0; i < aggregationInputTypes.size(); i++) {
+            finalPage[page.getChannelCount() + maskBlockChannels.length + i] = nullRle(aggregationInputTypes.get(i), positionCount);
+        }
+
+        // use raw input mask channel
+        finalPage[finalPage.length - 1] = nullRle(BOOLEAN, positionCount);
+        return new Page(positionCount, finalPage);
+    }
+
+    public Page processRawInputPage(Page page)
+    {
+        Block[] outputBlocks = new Block[hashChannels.length + groupedAggregators.size() + maskBlockChannels.length + aggregationInputChannels.length + 1];
+        int blockOffset = 0;
+        for (int i = 0; i < hashChannels.length; i++, blockOffset++) {
+            outputBlocks[blockOffset] = page.getBlock(hashChannels[i]);
+        }
+        for (int i = 0; i < groupedAggregators.size(); i++, blockOffset++) {
+            outputBlocks[blockOffset] = nullRle(groupedAggregators.get(i).getType(), page.getPositionCount());
+        }
+        for (int i = 0; i < maskBlockChannels.length; i++, blockOffset++) {
+            outputBlocks[blockOffset] = page.getBlock(maskBlockChannels[i]);
+        }
+        for (int i = 0; i < aggregationInputChannels.length; i++, blockOffset++) {
+            outputBlocks[blockOffset] = page.getBlock(aggregationInputChannels[i]);
+        }
+        outputBlocks[blockOffset] = booleanRle(true, page.getPositionCount());
+        return new Page(page.getPositionCount(), outputBlocks);
+    }
+
+    private static RunLengthEncodedBlock booleanRle(boolean value, int positionCount)
```

Comment:
There is a `RunLengthEncodedBlock.create` that accepts a single value and converts it to block